### PR TITLE
locale: initial add-on

### PR DESCRIPTION
**Pull request by thoradia**
I am building an image to test this.
The idea: if /storage/.config/locale is a directory then set the language accordingly.
If it works I will create an add-on to create UTF-8 locales.

### CODE REVIEW

**Comment by chewitt** [comment]
@thoradia the persistent storage location needs to be /storage/.kodi/addons/name-of-addon so that locale content is deleted if the user uninstalls the locale add-on.

**Comment by thoradia** [comment]
Rebased this as an add-on. Usage:

export LANG="it_IT@euro.UTF-8"
export LOCPATH="$(locpath)"

locale
LANG=it_IT@euro.UTF-8
LC_CTYPE="it_IT@euro.UTF-8"
LC_NUMERIC="it_IT@euro.UTF-8"
LC_TIME="it_IT@euro.UTF-8"
LC_COLLATE="it_IT@euro.UTF-8"
LC_MONETARY="it_IT@euro.UTF-8"
LC_MESSAGES="it_IT@euro.UTF-8"
LC_PAPER="it_IT@euro.UTF-8"
LC_NAME="it_IT@euro.UTF-8"
LC_ADDRESS="it_IT@euro.UTF-8"
LC_TELEPHONE="it_IT@euro.UTF-8"
LC_MEASUREMENT="it_IT@euro.UTF-8"
LC_IDENTIFICATION="it_IT@euro.UTF-8"
LC_ALL=

date
sab apr 15 16:38:24 CEST 2017

File: packages/addons/tools/locale/package.mk [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+#  You should have received a copy of the GNU General Public License
+#  along with LibreELEC.  If not, see <http://www.gnu.org/licenses/>.
+################################################################################
+
+PKG_NAME="locale"
+PKG_REV="100"
+PKG_ARCH="any"
+PKG_DEPENDS_TARGET="toolchain glibc"
+PKG_SECTION="virtual"
+PKG_SHORTDESC="Locales: compile and set a custom locale definition"
```
Comment (chewitt) [inline packages/addons/tools/locale/package.mk]: locale: allows users to set a custom locale to override the POSIX default

File: packages/addons/tools/locale/package.mk [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+PKG_LONGDESC="Locales ($PKG_REV) provides a script compile and set a custom locale definition"
```
Comment (chewitt) [inline packages/addons/tools/locale/package.mk]: locale ($PKG_REV) allows users to set a custom locale in the OS to override the POSIX default

File: packages/addons/tools/locale/source/bin/locpath [hunk 3]
```diff
@@ -0,0 +1,33 @@
+#!/bin/sh
+################################################################################
+#      This file is part of LibreELEC - http://www.libreelec.tv
```
Comment (chewitt) [inline packages/addons/tools/locale/source/bin/locpath]: https://libreelec.tv

File: packages/addons/tools/locale/source/default.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
+################################################################################
+#      This file is part of LibreELEC - http://www.libreelec.tv
```
Comment (chewitt) [inline packages/addons/tools/locale/source/default.py]: and again https://libreelec.tv

**Comment by chewitt** [comment]
@thoradia can you add this icon http://chewitt.libreelec.tv/images/icon_locale.png .. it's in our house style for official add-ons. Are you planning to add a simple GUI so users can configure the locale from within Kodi? ..and is the name "locale" or "virtual.locale" ? perhaps using a name like libreelec.locale would be best so the origin is clear when it shows up in file lists and logs.

**Comment by thoradia** [comment]
Thank you for the feedback @chewitt

The name is "locale", the section is "virtual", the full name, section.name, is virtual.locale. Users will see the add-on as Locale in Kodi. The add-on already has LibreELEC written all over it (in its maintainer, repository, requires, etc).

The add-on will mainly serve as a dependency of other add-ons or of third-party programs. I would not change the locale of LibreELEC (as opposed to changing the locale of a busybox subshell), as it may introduce unexpected behaviours (in sorts for example).

Finally, custom locales are now built in ADDON_DIR, rather than ADD_HOME, so that they are re-compiled if the add-on is updated.

**Comment by chewitt** [comment]
Handling sorts correctly and allowing file operations on non-latin files, e.g. Cyrillic or Hebrew is the main reason I was asking you to re-look at the add-on. If there are unintended consequences they should be found and fixed. I appreciate i18n can be a complex topic, but IMHO it's also part of being a grown-up distro serving a global audience :)

**Comment by thoradia** [comment]
I think we agree: this add-on allows for add-ons and third-party programs to handle non-latin files, but does not modify the locale of LibreELEC.

These two lines will only change the locale of a busybox subshell, not that of LibreELEC:
export LANG="it_IT@euro.UTF-8"
export LOCPATH="$(locpath)"

Which locale to use lies with the (users of) add-ons and third party program.

**Comment by chewitt** [comment]
I'd like others to vet the add-on before I merge, but long-term I would like it to be capable of changing the core OS locale - because that is what allows me (as a user) to SSH in and bulk-delete the awful stuff from my wife's Русский язык mp3 library. Once this add-on is installed I can export ru_RU@UTF-8 for the session, but if I am a native Cyrillic user and want this to be the default the goal is to set locale via the GUI without needing to hack the console. However, "Rome wasn't built in a day" so something for v101 in the future, maybe.

**Comment by thoradia** [comment]
echo "export LANG=ru_RU.UTF-8; export LOCPATH=$(locpath)" > russian
. russian

I will add a setting to select a default language from the add-on, and a corresponding command to set LANG accordingly.

**Comment by chewitt** [comment]
btw, I'm sure this can/will break something but the goal is to find out what/where - that's what I meant in the previous comment. Thanks :)

File: packages/sysutils/busybox/profile.d/98-busybox.conf [hunk 5]
```diff
@@ -32,3 +32,7 @@ case "$TERM" in
     export TERM="xterm"
     ;;
 esac
+if  $(which loc_busybox) && loc_busybox; then
+  export LANG="$(loc_locale)"
+  export LOCPATH="$(loc_path)"
+fi
```
Comment (thoradia) [inline packages/sysutils/busybox/profile.d/98-busybox.conf]: Thnak you for the hint

File: packages/addons/tools/locale/source/resources/settings.xml [hunk 6]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8" standalone="yes"?>
+<settings>
+   <category   label="30000">
+      <setting label="30001" id="loc_locale"  type="select" default="en_US" values="en_US|ru_RU" />
```
Comment (zbyna) [inline packages/addons/tools/locale/source/resources/settings.xml]: Is it possible to add cs_CZ  please ?

**Comment by thoradia** [comment]
This sort of works.
To-dos:
1. create a list of locales from files in i18n/locale
2. reboot when locale changes

Help welcome

**Comment by zbyna** [comment]
Not sure if it helps, but enclose list of locales (json string) from usr/share/i18n/locale:
[locales_list.txt](https://github.com/LibreELEC/LibreELEC.tv/files/965552/locales_list.txt)

**Comment by thoradia** [comment]
Nearly there...

**Comment by thoradia** [comment]
I think I am done.
Please test.

Note that LANG does not appear to affect Kodi.

**Comment by thoradia** [comment]
# locale
LANG=ja_JP.UTF-8
LC_CTYPE="ja_JP.UTF-8"
LC_NUMERIC="ja_JP.UTF-8"
LC_TIME="ja_JP.UTF-8"
LC_COLLATE="ja_JP.UTF-8"
LC_MONETARY="ja_JP.UTF-8"
LC_MESSAGES="ja_JP.UTF-8"
LC_PAPER="ja_JP.UTF-8"
LC_NAME="ja_JP.UTF-8"
LC_ADDRESS="ja_JP.UTF-8"
LC_TELEPHONE="ja_JP.UTF-8"
LC_MEASUREMENT="ja_JP.UTF-8"
LC_IDENTIFICATION="ja_JP.UTF-8"
LC_ALL=
# date
木  5月  4 14:42:05 CEST 2017

**Comment by vpeter4** [comment]
Why id this addon declared as service? Because no extra service is running from it.

**Comment by thoradia** [comment]
This add-on is declared as a service to run default.py.
default.py creates locale and profile on start-up and on settings changed.

**Comment by thoradia** [comment]
This has been running without a problem on my system for four days.
@chewitt does it work for you?

**Comment by chewitt** [comment]
@thoradia I'll test in the coming days - need to get back from Kodi DevCon first :)

**Comment by chewitt** [comment]
I see this when running `locale` from the console:

```Play2:~ # locale
locale: Cannot set LC_CTYPE to default locale: No such file or directory
locale: Cannot set LC_MESSAGES to default locale: No such file or directory
locale: Cannot set LC_ALL to default locale: No such file or directory
LANG=en_GB.UTF-8
LC_CTYPE="en_GB.UTF-8"
LC_NUMERIC="en_GB.UTF-8"
LC_TIME="en_GB.UTF-8"
LC_COLLATE="en_GB.UTF-8"
LC_MONETARY="en_GB.UTF-8"
LC_MESSAGES="en_GB.UTF-8"
LC_PAPER="en_GB.UTF-8"
LC_NAME="en_GB.UTF-8"
LC_ADDRESS="en_GB.UTF-8"
LC_TELEPHONE="en_GB.UTF-8"
LC_MEASUREMENT="en_GB.UTF-8"
LC_IDENTIFICATION="en_GB.UTF-8"
LC_ALL=```

I'm not sure if that's important or not?

File: packages/addons/service/locale/source/settings-default.xml [hunk 7]
```diff
@@ -0,0 +1,4 @@
+<settings>
+    <setting id="charmap" value="UTF-8" />
+    <setting id="locale" value="en_US" />
```
Comment (chewitt) [inline packages/addons/service/locale/source/settings-default.xml]: Kodi default GUI language is en_GB not en_US

File: packages/addons/service/locale/source/resources/language/English/strings.po [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+msgctxt "#30001"
+msgid "Charmap"
+msgstr ""
+
+msgctxt "#30002"
+msgid "Locale"
+msgstr ""
+
+msgctxt "#30003"
+msgid "LANG has changed from {} to {}."
```
Comment (chewitt) [inline packages/addons/service/locale/source/resources/language/English/strings.po]: You can put "Locale changed from {} to {}. Please reboot to apply globally." in a single msgid. I'm not sure there's a reason to split the statements into two.

**Comment by vpeter4** [comment]
> I see this when running locale from the console:

Which LibreELEC version? On 8.0.1 I don't see that.

**Comment by thoradia** [comment]
@chewitt 
I have made the recommended changes.
I can not reproduce the warnings you see, neither on install nor on update.

**Comment by chewitt** [comment]
@vpeter4 I'm using a pre-release 8.0.2 image from the usual team/my share location

**Comment by vpeter4** [comment]
@chewitt: I'm testing in Virtual and don't see messages you got. Seems to be working fine for me.

**Comment by thoradia** [comment]
@chewitt does the issue persist if you uninstall then reinstall the add-on?
Not sure how to debug without a WP2

**Comment by thoradia** [comment]
I will add this to my tree (with a low revision number) until it is merged in LibreELEC.

**Comment by chewitt** [comment]
I noticed when poking around ~ there is no content in the en_GB.UTF-8 folder but there is for the ru_RU.UTF-8 folder: ```Play2:~ # cd /storage/.kodi/addons/service.locale/locpath/en_GB.UTF-8/
Play2:~/.kodi/addons/service.locale/locpath/en_GB.UTF-8 # ls -l
total 0
Play2:~/.kodi/addons/service.locale/locpath/en_GB.UTF-8 # cd ../ru_RU.UTF-8/
Play2:~/.kodi/addons/service.locale/locpath/ru_RU.UTF-8 # ls -l
total 1538
-rw-r--r--    1 root     root           151 May 15 14:40 LC_ADDRESS
-rw-r--r--    1 root     root       1243770 May 15 14:40 LC_COLLATE
-rw-r--r--    1 root     root        317872 May 15 14:40 LC_CTYPE
-rw-r--r--    1 root     root           341 May 15 14:40 LC_IDENTIFICATION
-rw-r--r--    1 root     root            23 May 15 14:40 LC_MEASUREMENT
drwxr-xr-x    2 root     root          1024 May 15 14:40 LC_MESSAGES
-rw-r--r--    1 root     root           294 May 15 14:40 LC_MONETARY
-rw-r--r--    1 root     root            62 May 15 14:40 LC_NAME
-rw-r--r--    1 root     root            54 May 15 14:40 LC_NUMERIC
-rw-r--r--    1 root     root            34 May 15 14:40 LC_PAPER
-rw-r--r--    1 root     root            52 May 15 14:40 LC_TELEPHONE
-rw-r--r--    1 root     root          2498 May 15 14:40 LC_TIME```

Do you see different on your devices?

**Comment by chewitt** [comment]
further info:

```Play2:~/.kodi/addons/service.locale/locpath # cp -R en_US.UTF-8/* en_GB.UTF-8/
Play2:~/.kodi/addons/service.locale/locpath # locale
LANG=en_GB.UTF-8
LC_CTYPE="en_GB.UTF-8"
LC_NUMERIC="en_GB.UTF-8"
LC_TIME="en_GB.UTF-8"
LC_COLLATE="en_GB.UTF-8"
LC_MONETARY="en_GB.UTF-8"
LC_MESSAGES="en_GB.UTF-8"
LC_PAPER="en_GB.UTF-8"
LC_NAME="en_GB.UTF-8"
LC_ADDRESS="en_GB.UTF-8"
LC_TELEPHONE="en_GB.UTF-8"
LC_MEASUREMENT="en_GB.UTF-8"
LC_IDENTIFICATION="en_GB.UTF-8"
LC_ALL=```

**Comment by chewitt** [comment]
So I rebuilt the add-on after a clean build run and I no longer see the error so I guess this was a dirty build dirs "my bad" issue ~ sorry for the noise. The only nit-pick issue I can find now is that "POSIX" doesn't show in the before/after message when the add-on is first installed and run and the locale changes from POSIX to en_GB.UTF-8, see below: 

screenshot000

It's a minor cosmetic so I'm happy to merge the add-on, but it would be nice to fix that up at some point. Thanks for the contribution :)

**Pull request merged**